datanode/api/trading_data_v2.go

```diff
@@ -2647,16 +2659,205 @@ func (t *TradingDataServiceV2) estimateMargin(
+		var err error
+		collateralAvailable, err = num.DecimalFromString(*req.CollateralAvailable)
+		if err != nil {
+			return nil, formatE(ErrPositionsInvalidCollateralAmount, err)
+		}
+	}
+
+	mkt, err := t.marketService.GetByID(ctx, req.MarketId)
+	if err != nil {
+		return nil, formatE(ErrMarketServiceGetByID, err)
+	}
+
+	priceFactor, err := t.getMarketPriceFactor(ctx, mkt)
+	if err != nil {
+		return nil, err
+	}
+
+	dPriceFactor := priceFactor.ToDecimal()
+
+	buyOrders := make([]*risk.OrderInfo, 0, len(req.Orders))
+	sellOrders := make([]*risk.OrderInfo, 0, len(req.Orders))
+
+	for _, o := range req.Orders {
+		if o == nil {
+			continue
+		}
+		var price num.Decimal
+		p, err := num.DecimalFromString(o.Price)
+		if err != nil {
+			return nil, errors.Wrap(ErrInvalidOrderPrice, err.Error())
+		}
+		price = p.Mul(dPriceFactor)
+
+		if o.Side == types.SideBuy {
+			buyOrders = append(buyOrders, &risk.OrderInfo{Size: o.Remaining, Price: price, IsMarketOrder: o.IsMarketOrder})
+		}
+		if o.Side == types.SideSell {
+			sellOrders = append(sellOrders, &risk.OrderInfo{Size: o.Remaining, Price: price, IsMarketOrder: o.IsMarketOrder})
+		}
+	}
+
+	rf, err := t.riskFactorService.GetMarketRiskFactors(ctx, req.MarketId)
+	if err != nil {
+		return nil, formatE(ErrRiskFactorServiceGet, err)
+	}
+
+	mktData, err := t.marketDataService.GetMarketDataByID(ctx, req.MarketId)
+	if err != nil {
+		return nil, formatE(ErrMarketServiceGetMarketData, err)
+	}
+
+	mktProto := mkt.ToProto()
+
+	asset, err := mktProto.GetAsset()
+	if err != nil {
+		return nil, formatE(err)
+	}
+
+	marketObservable, err := num.DecimalFromString(mktData.MarkPrice.String())
+	if err != nil {
+		return nil, formatE(err)
+	}
+
+	auction := mktData.AuctionEnd > 0
+	if auction && mktData.MarketTradingMode == types.MarketTradingModeOpeningAuction.String() {
+		marketObservable = mktData.IndicativePrice
+	}
+
+	uMarketObservable, overflowed := num.UintFromDecimal(marketObservable)
+	if overflowed {
+		return nil, formatE(fmt.Errorf("mark price overflowed: %s", mktData.MarkPrice))
+	}
+
+	marketObservableInAssetPrecision, err := t.scaleFromMarketToAssetPrice(ctx, mkt, uMarketObservable)
+	if err != nil {
+		return nil, formatE(ErrScalingPriceFromMarketToAsset, err)
+	}
+
+	dMarketObservableInAssetPrecision := num.DecimalFromUint(marketObservableInAssetPrecision)
+
+	positionFactor := num.DecimalFromFloat(10).
+		Pow(num.DecimalFromInt64(int64(mkt.PositionDecimalPlaces)))
+
+	linearSlippageFactor, err := num.DecimalFromString(mktProto.LinearSlippageFactor)
+	if err != nil {
+		return nil, formatE(fmt.Errorf("can't parse linear slippage factor: %s", mktProto.LinearSlippageFactor), err)
+	}
+	quadraticSlippageFactor, err := num.DecimalFromString(mktProto.QuadraticSlippageFactor)
+	if err != nil {
+		return nil, formatE(fmt.Errorf("can't parse quadratic slippage factor: %s", mktProto.QuadraticSlippageFactor), err)
+	}
+
+	marginEstimate := t.computeMarginRange(
```

Comment:
A function with the name `computeMarginRange`, and in general this much processing/logic really doesn't shouldn't be sitting in a handler like this. Convert the input required, and pass it to a component that contains the meat of the logic/work that needs to be done. It makes it easier to review, maintain, and work on

Reply:
Did you look inside the function? It just calls an external component twice, so not sure what you mean?